tests/runtime_tests/test_finite_differences.py

```diff
@@ -0,0 +1,151 @@
+
+    class InputSchema(BaseModel):
+        in_data: Differentiable[Array[(None, 3), Float32]]
+        in_dict: dict[str, Differentiable[Array[(3, 3, 3), Float32]]]
+        in_scalar: Differentiable[Float32]
+        in_aux: str
+
+    class OutputSchema(BaseModel):
+        out_data: Differentiable[Array[(None, 3), Float32]]
+        out_dict: dict[str, Differentiable[Array[(3, 3, 3), Float32]]]
+        out_scalar: Differentiable[Float32]
+        out_aux: str
+
+    def apply(inputs: InputSchema) -> OutputSchema:
+        return {
+            "out_data": np.zeros_like(inputs.in_data),
+            "out_dict": {
+                key: np.zeros_like(value) for key, value in inputs.in_dict.items()
+            },
+            "out_scalar": np.zeros_like(inputs.in_scalar),
+            "out_aux": inputs.in_aux,
+        }
+
+    def jacobian(
+        inputs: InputSchema,
+        jac_inputs: set[str],
+        jac_outputs: set[str],
+    ):
+        outputs = DummyModule.apply(inputs)
+        if DummyModule.correct_gradients:
+            make_array = np.zeros
+        else:
+            make_array = np.ones
+
+        return {
+            key_out: {
+                key_in: make_array(
+                    (
+                        *get_at_path(outputs, key_out).shape,
+                        *get_at_path(inputs, key_in).shape,
+                    )
+                )
+                for key_in in jac_inputs
+            }
+            for key_out in jac_outputs
+        }
+
+    def jacobian_vector_product(
+        inputs: InputSchema,
+        jac_inputs: set[str],
+        jac_outputs: set[str],
+        vector: Array[(None,), Float32],
+    ):
+        outputs = DummyModule.apply(inputs)
+        if DummyModule.correct_gradients:
+            make_array = np.zeros
+        else:
+            make_array = np.ones
+        return {
+            key_out: make_array(get_at_path(outputs, key_out).shape)
+            for key_out in jac_outputs
+        }
+
+    def vector_jacobian_product(
+        inputs: InputSchema,
+        jac_inputs: set[str],
+        jac_outputs: set[str],
+        vector: Array[(None,), Float32],
+    ):
+        if DummyModule.correct_gradients:
+            make_array = np.zeros
+        else:
+            make_array = np.ones
+        return {
+            key_in: make_array(get_at_path(inputs, key_in).shape)
+            for key_in in jac_inputs
+        }
+
+
+rng = np.random.default_rng(0)
+input_data = {
+    "in_data": rng.random((10, 3)),
+    "in_dict": {"key": rng.random((3, 3, 3))},
+    "in_scalar": rng.random(1)[0],
+    "in_aux": "auxiliary",
+}
+
+
+@pytest.mark.parametrize("input_paths", [None, ["in_data"], ["in_dict.{key}"]])
+@pytest.mark.parametrize("output_paths", [None, ["out_data", "out_dict.{key}"]])
+@pytest.mark.parametrize("endpoints", [None, ["jacobian"]])
+def test_check_gradients(input_paths, output_paths, endpoints):
+    with typeguard.suppress_type_checks():
```

Comment:
Had to do this because `DummyModule` is `ModuleType`, not an actual module – will have another look at this